astropy/table/tests/test_column.py

```diff
@@ -583,3 +584,142 @@ def test_string_truncation_warning():
         assert len(w) == 1
         assert ('truncated right side string(s) longer than 2 character(s)'
                 in str(w[0].message))
+
+
+@pytest.mark.skipif('six.PY2')
+@pytest.mark.parametrize('Column', (table.Column, table.MaskedColumn))
+def test_col_unicode_sandwich_create_from_str(Column):
+    """
+    Create a bytestring Column from strings (including unicode) in Py3.
+    """
+    # a-umlaut is a 2-byte character in utf-8, test fails with ascii encoding.
+    # Stress the system by injecting non-ASCII characters.
+    uba = u'bä'
+    c = Column([uba, 'def'], dtype='S')
+    assert c.dtype.char == 'S'
+    assert c[0] == uba
+    assert isinstance(c[0], str)
+    assert isinstance(c[:0], table.Column)
+    assert np.all(c[:2] == np.array([uba, 'def']))
+
+
+@pytest.mark.parametrize('Column', (table.Column, table.MaskedColumn))
+def test_col_unicode_sandwich_bytes(Column):
+    """
+    Create a bytestring Column from bytes and ensure that it works in Python 3 in
+    a convenient way like in Python 2.
+    """
+    # a-umlaut is a 2-byte character in utf-8, test fails with ascii encoding.
+    # Stress the system by injecting non-ASCII characters.
+    uba = 'ba' if six.PY2 else u'bä'
+    uba8 = uba.encode('utf-8')
+    c = Column([uba8, b'def'])
+    assert c.dtype.char == 'S'
+    assert c[0] == uba8 if six.PY2 else uba  # Can compare utf-8 directly only in PY3
+    assert isinstance(c[0], str)
+    assert isinstance(c[:0], table.Column)
+    assert np.all(c[:2] == np.array([uba, 'def']))
+
+    assert isinstance(c[:], table.Column)
+    assert c[:].dtype.char == 'S'
+
+    # Array / list comparisons
+    if not six.PY2:
+        assert np.all(c == [uba, 'def'])
+
+    ok = c == [uba8, b'def']
+    assert type(ok) is type(c.data)
+    assert ok.dtype.char == '?'
+    assert np.all(ok)
+
+    assert np.all(c == np.array([uba, u'def']))
+    if not six.PY2:
+        assert np.all(c == np.array([uba8, b'def']))
+
+    # Scalar compare
+    cmps = (uba8,) if six.PY2 else (uba, uba8)
+    for cmp in cmps:
+        ok = c == cmp
+        assert type(ok) is type(c.data)
+        assert np.all(ok == [True, False])
+
+
+def test_col_unicode_sandwich_unicode():
+    """
+    Sanity check that Unicode Column behaves normally.
+    """
+    uba = 'ba' if six.PY2 else u'bä'
```

Comment:
In this case the final test fails, for essentially this reason:
```
In [2]: uba = u'bä'
In [4]: uba8 = uba.encode('utf-8')
In [5]: np.array(uba) == np.array(uba8)
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-5-c238e27a518e> in <module>()
----> 1 np.array(uba) == np.array(uba8)

UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 1: ordinal not in range(128)
```
So I either leave this alone or remove the final test.  Preference?